src/Symfony/Component/Security/Core/Role/RoleHierarchyInterface.php

```diff
@@ -29,6 +29,8 @@ interface RoleHierarchyInterface
      * @param Role[] $roles An array of directly assigned roles
      *
      * @return Role[] An array of all reachable roles
+     *
+     * @deprecated since Symfony 4.3, to be removed in 5.0.
```

Comment:
There is no upgrade path for deprecating a single method in an interface, right? I mean it's impossible for an implementation to be deprecation free before 5.0, isn't it?

Reply:
sure there is: we can remove the method here

Reply:
Not implementing the deprecated method would mean that code may break other packages that still rely on the legacy interface.

Reply:
The deprecation policy is about already written code, so we don't really care here.

Reply:
I think we should. If I install package X in my application that uses the 4.3 API and I install another package Y which still makes use of the 4.2 API things will break in an unexpected way. Both packages on their own act in a sane way, but the combination will not work anymore.

Reply:
How would that break? The new package would require security-core 4.3, so things would continue to work to me.

Reply:
Only package X requires 4.3. Package Y does not require 4.3 and still calls the old method, but that method does not exist anymore.

Reply:
The "no upgrade path" issue is a real one to me. We always seek for a way to be deprecation-free before moving to next major. I don't think we should care about hypothetical future code. If you really think we cannot remove the method, can we think of any other plan? If not, let's remove?

I think you have a point, but that allowing ppl to be deprecation free (providing a continuous upgrade path) is much more important than forward-compat with unwritten code.

What do others think?

Reply:
I agree that providing no upgrade path feels more critical.

Reply:
done

Reply:
Removing a method from an interface is a BC break. It causes exceptions such as:

> Prophecy\Exception\Doubler\MethodNotFoundException: Method `Double\RoleHierarchyInterface\P19::getReachableRoles()` is not defined.